man/Geod.1
.\" Automatically generated by Pod::Man 2.25 (Pod::Simple 3.16)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "GEOD 1"
.TH GEOD 1 "2013-01-16" "GeographicLib 1.29" "GeographicLib Utilities"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
Geod \-\- perform geodesic calculations
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBGeod\fR [ \fB\-i\fR | \fB\-l\fR \fIlat1\fR \fIlon1\fR \fIazi1\fR ] [ \fB\-a\fR ] [ \fB\-e\fR \fIa\fR \fIf\fR ]
[ \fB\-d\fR | \fB\-:\fR ] [ \fB\-b\fR ] [ \fB\-f\fR ] [ \fB\-p\fR \fIprec\fR ] [ \fB\-E\fR ]
[ \fB\-\-comment\-delimiter\fR \fIcommentdelim\fR ]
[ \fB\-\-version\fR | \fB\-h\fR | \fB\-\-help\fR ]
[ \fB\-\-input\-file\fR \fIinfile\fR | \fB\-\-input\-string\fR \fIinstring\fR ]
[ \fB\-\-line\-separator\fR \fIlinesep\fR ]
[ \fB\-\-output\-file\fR \fIoutfile\fR ]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The shortest path between two points on the ellipsoid at (\fIlat1\fR,
\&\fIlon1\fR) and (\fIlat2\fR, \fIlon2\fR) is called the geodesic.  Its length is
\&\fIs12\fR and the geodesic from point 1 to point 2 has forward azimuths
\&\fIazi1\fR and \fIazi2\fR at the two end points.
.PP
\&\fBGeod\fR operates in one of three modes:
.IP "1." 4
By default, \fBGeod\fR accepts lines on the standard input containing
\&\fIlat1\fR \fIlon1\fR \fIazi1\fR \fIs12\fR and prints \fIlat2\fR \fIlon2\fR \fIazi2\fR
on standard output.  This is the direct geodesic calculation.
.IP "2." 4
Command line arguments \fB\-l\fR \fIlat1\fR \fIlon1\fR \fIazi1\fR specify a geodesic line.
\&\fBGeod\fR then accepts a sequence of \fIs12\fR values (one per line) on
standard input and prints \fIlat2\fR \fIlon2\fR \fIazi2\fR for each.  This
generates a sequence of points on a single geodesic.
.IP "3." 4
With the \fB\-i\fR command line argument, \fBGeod\fR performs the inverse
geodesic calculation.  It reads lines containing \fIlat1\fR \fIlon1\fR \fIlat2\fR
\&\fIlon2\fR and prints the corresponding values of \fIazi1\fR \fIazi2\fR \fIs12\fR.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-i\fR" 4
.IX Item "-i"
perform an inverse geodesic calculation (see 3 above).
.IP "\fB\-l\fR" 4
.IX Item "-l"
line mode (see 2 above); generate a sequence of points along the
geodesic specified by \fIlat1\fR \fIlon1\fR \fIazi1\fR.
.IP "\fB\-a\fR" 4
.IX Item "-a"
arc mode; on input \fIand\fR output \fIs12\fR is replaced by \fIa12\fR the arc
length (in degrees) on the auxiliary sphere.  See \*(L"\s-1AUXILIARY\s0 \s-1SPHERE\s0\*(R".
.IP "\fB\-e\fR" 4
.IX Item "-e"
specify the ellipsoid via \fIa\fR \fIf\fR; the equatorial radius is \fIa\fR and
the flattening is \fIf\fR.  Setting \fIf\fR = 0 results in a sphere.  Specify
\&\fIf\fR < 0 for a prolate ellipsoid.  A simple fraction, e.g., 1/297,
is allowed for \fIf\fR.  (Also, if \fIf\fR > 1, the flattening is set to
1/\fIf\fR.)  By default, the \s-1WGS84\s0 ellipsoid is used, \fIa\fR = 6378137 m,
\&\fIf\fR = 1/298.257223563.
.IP "\fB\-d\fR" 4
.IX Item "-d"
output angles as degrees, minutes, seconds instead of decimal degrees.
.IP "\fB\-:\fR" 4
.IX Item "-:"
like \fB\-d\fR, except use : as a separator instead of the d, ', and "
delimiters.
.IP "\fB\-b\fR" 4
.IX Item "-b"
report the \fIback\fR azimuth at point 2 instead of the forward azimuth.
.IP "\fB\-f\fR" 4
.IX Item "-f"
full output; each line of output consists of 12 quantities: \fIlat1\fR
\&\fIlon1\fR \fIazi1\fR \fIlat2\fR \fIlon2\fR \fIazi2\fR \fIs12\fR \fIa12\fR \fIm12\fR \fIM12\fR
\&\fIM21\fR \fIS12\fR.  \fIa12\fR is described in \*(L"\s-1AUXILIARY\s0 \s-1SPHERE\s0\*(R".  The four
quantities \fIm12\fR, \fIM12\fR, \fIM21\fR, and \fIS12\fR are described in
\&\*(L"\s-1ADDITIONAL\s0 \s-1QUANTITIES\s0\*(R".
.IP "\fB\-p\fR" 4
.IX Item "-p"
set the output precision to \fIprec\fR (default 3); \fIprec\fR is the
precision relative to 1 m.  See \s-1PRECISION\s0.
.IP "\fB\-E\fR" 4
.IX Item "-E"
use \*(L"exact\*(R" algorithms (based on elliptic integrals) for the geodesic
calculations.  These are more accurate than the (default) series
expansions for |\fIf\fR| > 0.02.
.IP "\fB\-\-comment\-delimiter\fR" 4
.IX Item "--comment-delimiter"
set the comment delimiter to \fIcommentdelim\fR (e.g., \*(L"#\*(R" or \*(L"//\*(R").  If
set, the input lines will be scanned for this delimiter and, if found,
the delimiter and the rest of the line will be removed prior to
processing and subsequently appended to the output line (separated by a
space).
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
print version and exit.
.IP "\fB\-h\fR" 4
.IX Item "-h"
print usage and exit.
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
print full documentation and exit.
.IP "\fB\-\-input\-file\fR" 4
.IX Item "--input-file"
read input from the file \fIinfile\fR instead of from standard input; a file
name of \*(L"\-\*(R" stands for standard input.
.IP "\fB\-\-input\-string\fR" 4
.IX Item "--input-string"
read input from the string \fIinstring\fR instead of from standard input.
All occurrences of the line separator character (default is a semicolon)
in \fIinstring\fR are converted to newlines before the reading begins.
.IP "\fB\-\-line\-separator\fR" 4
.IX Item "--line-separator"
set the line separator character to \fIlinesep\fR.  By default this is a
semicolon.
.IP "\fB\-\-output\-file\fR" 4
.IX Item "--output-file"
write output to the file \fIoutfile\fR instead of to standard output; a
file name of \*(L"\-\*(R" stands for standard output.
.SH "INPUT"
.IX Header "INPUT"
\&\fBGeod\fR measures all angles in degrees and all lengths (\fIs12\fR) in
meters.  On input angles (latitude, longitude, azimuth, arc length) can
be as decimal degrees or degrees (d), minutes ('), seconds (\*(L").  A
decimal point can only appear in the least significant component and the
designator (d, ', or \*(R") for this component is optional; thus \f(CW\*(C`40d30\*(C'\fR,
\&\f(CW\*(C`40d30\*(Aq\*(C'\fR, \f(CW\*(C`40.5d\*(C'\fR, and \f(CW40.5\fR are all equivalent.  By default,
latitude precedes longitude for each point; however on input either may
be given first by appending (or prepending) \fIN\fR or \fIS\fR to the latitude
and \fIE\fR or \fIW\fR to the longitude.  Azimuths are measured clockwise from
north; however this may be overridden with \fIE\fR or \fIW\fR.
.PP
See the \f(CW\*(C`QUOTING\*(C'\fR section of \fIGeoConvert\fR\|(1) for how to quote the \s-1DMS\s0
designators ' and ".
.SH "AUXILIARY SPHERE"
.IX Header "AUXILIARY SPHERE"
Geodesics on the ellipsoid can be transferred to the \fIauxiliary sphere\fR
on which the distance is measured in terms of the arc length \fIa12\fR
(measured in degrees) instead of \fIs12\fR.  In terms of \fIa12\fR, 180
degrees is the distance from one equator crossing to the next or from
the minimum latitude to the maximum latitude.  Geodesics with \fIa12\fR
> 180 degrees do not correspond to shortest paths.  With the \fB\-a\fR
flag, \fIs12\fR (on both input and output) is replaced by \fIa12\fR.  The
\&\fB\-a\fR flag does \fInot\fR affect the full output given by the \fB\-f\fR flag
(which always includes both \fIs12\fR and \fIa12\fR).
.SH "ADDITIONAL QUANTITIES"
.IX Header "ADDITIONAL QUANTITIES"
The \fB\-f\fR flag reports four additional quantities.
.PP
The reduced length of the geodesic, \fIm12\fR, is defined such that if the
initial azimuth is perturbed by d\fIazi1\fR (radians) then the second point
is displaced by \fIm12\fR d\fIazi1\fR in the direction perpendicular to the
geodesic.  \fIm12\fR is given in meters.  On a curved surface the
reduced length obeys a symmetry relation, \fIm12\fR + \fIm21\fR = 0.  On a
flat surface, we have \fIm12\fR = \fIs12\fR.
.PP
\&\fIM12\fR and \fIM21\fR are geodesic scales.  If two geodesics are parallel at
point 1 and separated by a small distance \fIdt\fR, then they are separated
by a distance \fIM12\fR \fIdt\fR at point 2.  \fIM21\fR is defined similarly
(with the geodesics being parallel to one another at point 2).  \fIM12\fR
and \fIM21\fR are dimensionless quantities.  On a flat surface, we have
\&\fIM12\fR = \fIM21\fR = 1.
.PP
If points 1, 2, and 3 lie on a single geodesic, then the following
addition rules hold,
\&\fIs13\fR = \fIs12\fR + \fIs23\fR,
\&\fIa13\fR = \fIa12\fR + \fIa23\fR,
\&\fIS13\fR = \fIS12\fR + \fIS23\fR,
\&\fIm13\fR = \fIm12\fR \fIM23\fR + \fIm23\fR \fIM21\fR,
\&\fIM13\fR = \fIM12\fR \fIM23\fR \- (1 \- \fIM12\fR \fIM21\fR) \fIm23\fR / \fIm12\fR, and
\&\fIM31\fR = \fIM32\fR \fIM21\fR \- (1 \- \fIM23\fR \fIM32\fR) \fIm12\fR / \fIm23\fR.
.PP
Finally, \fIS12\fR is the area between the geodesic from point 1 to point 2
and the equator; i.e., it is the area, measured counter-clockwise, of
the quadrilateral with corners (\fIlat1\fR,\fIlon1\fR), (0,\fIlon1\fR),

NEWS
A reverse chronological list of changes to GeographicLib

For more information, see

    http://geographiclib.sourceforge.net/

The current version of the library is 1.29.

Changes between 1.29 (released 2013-01-16) and 1.28 versions:

   * Changes to allow compilation with libc++ (courtesy of Kal Conley).

   * Add description of geodesics on triaxial ellipsoid to
     documentation.

   * Update journal reference for "Algorithms for geodesics".

Changes between 1.28 (released 2012-12-11) and 1.27 versions:

   * Changes to geodesic routines:
     + compute longitude difference exactly;
     + hence fix BUG in area calculations for polygons with vertices very
       close to the prime meridian;
     + fix BUG is geoddistance.m where the value of m12 was wrong for
       meridional geodesics;
     + add Matlab implementations of the geodesic projections;
     + remove unneeded special code for geodesics which start at a pole;
     + include polygon area routine in C and Fortran implementations;
     + add doxygen documentation for C and Fortran libraries.

Changes between 1.27 (released 2012-11-29) and 1.26 versions:

   * Changes to geodesic routines:
     + add native Matlab implementations: geoddistance.m, geodreckon.m,
       geodarea.m;
     + add C and Fortran implementations;
     + improve the solution of the direct problem so that the series
       solution is accurate to round off for |f| < 1/50;
     + tighten up the convergence criteria for solution of the inverse
       problem;
     + no longer signal failures of convergence with NaNs (a slightly
       less accurate answer is returned instead).

   * Fix DMS::Decode double rounding BUG.

   * On MacOSX platforms with the cmake configuration, universal
     binaries are built.

Changes between 1.26 (released 2012-10-22) and 1.25 versions:

   * Replace the series used for geodesic areas by one with better
     convergence (this only makes an appreciable difference if |f| >
     1/150).

Changes between 1.25 (released 2012-10-16) and 1.24 versions:

   * Changes to geodesic calculations:
     + restart Newton's method in Geodesic::Inverse when it goes awry;
     + back up Newton's method with the bisection method;
     + Geodesic::Inverse now converges for any value of f;
     + add GeodesicExact and GeodesicLineExact which are formulated in
       terms of elliptic integrals and thus yield accurate results
       even for very eccentric ellipsoids.
     + the -E option to Geod invokes these exact classes.

   * Add functionality to EllipticFunction:
     + add all the traditional elliptic integrals;
     + remove restrictions on argument range for incomplete elliptic
       integrals;
     + allow imaginary modulus for elliptic integrals and elliptic
       functions;
     + make interface to the symmetric elliptic integrals public.

   * Allow GeographicLib::Ellipsoid to be copied.

   * Changes to the build tools:
     + cmake uses folders in Visual Studio to reduce clutter;
     + allow precision of reals to be set in cmake;
     + fail gracefully in the absence of pod documentation tools;
     + remove support for maintainer tasks in Makefile.mk.

Changes between 1.24 (released 2012-09-22) and 1.23 versions:

   * Allow the specification of the hemisphere in UTM coordinates in
     order to provide continuity across the equator:
     + add UTMUPS::Transfer;
     + add GeoCoords::UTMUPSRepresentation(bool, int) and
       GeoCoords::AltUTMUPSRepresentation(bool, int);
     + use the hemisphere letter in, e.g., GeoConvert -u -z 31N.

   * Add UTMUPS::DecodeEPSG and UTMUPS::EncodeEPSG.

   * cmake changes:
     + restore support for cmake 2.4.x;
     + explicitly check version of doxygen.

   * Fix building under cygwin.

   * Document restrictions on f in the Introduction.

   * Fix python interface to work with version 2.6.x.

Changes between 1.23 (released 2012-07-17) and 1.22 versions:

  * Documentation changes:
    + remove html documentation from distribution and use web links if
      doxygen is not available;
    + use doxygen tags to document exceptions;
    + begin migrating the documentation to using Greek letters where
      appropriate (requires doxygen 1.8.1.2 or later).

  * Add Math::AngNormalize and Math::AngNormalize2; the allowed range
    for longitudes and azimuths widened to [-540d, 540d).

  * DMS::Decode understands more unicode symbols.

  * Geohash uses geohash code "nan" to stand for not a number.

  * Add Ellipsoid::NormalCurvatureRadius.

  * Various fixes in LambertConformalConic, TransverseMercator,
    PolarStereographic, and Ellipsoid to handle reverse projections of
    points near infinity.

  * Fix programming blunder in LambertConformalConic::Forward (incorrect
    results were returned if the tangent latitude was negative).

Changes between 1.22 (released 2012-05-27) and 1.21 versions:

  * Add Geohash and Ellipsoid classes.

  * Fix bug in AlbersEqualArea of very prolate ellipsoids (b^2 > 2
    a^2).

  * cmake changes:
    + optionally use PACKAGE_PATH and INSTALL_PATH to determine
      CMAKE_INSTALL_PREFIX;
    + use COMMON_INSTALL_PATH to determine layout of installation
      directories;
    + as a consequence, the installation paths for the documentation,
      and python and matlab interfaces are shortened for Windows;
    + zip source distribution now uses DOS line endings;
    + the tests work in debug mode for Windows;
    + default setting of GEOGRAPHICLIB_DATA does not depend on
      CMAKE_INSTALL_PREFIX;
    + add a cmake configuration for build tree.

Changes between 1.21 (released 2012-04-25) and 1.20 versions:

  * Support colon-separated DMS output:
    + DMS::Encode and GeoCoords::DMSRepresentation generalized;
    + GeoConvert and Geod now accept a -: option.

  * GeoidEval does not print the gradient of the geoid height by default
    (because it's subject to large errors); give the -g option to get
    the gradient printed.


tools/Geod.cpp
/**
 * \file Geod.cpp
 * \brief Command line utility for geodesic calculations
 *
 * Copyright (c) Charles Karney (2009-2012) <charles@karney.com> and licensed
 * under the MIT/X11 License.  For more information, see
 * http://geographiclib.sourceforge.net/
 *
 * Compile and link with
 *   g++ -g -O3 -I../include -I../man -o Geod \
 *       Geod.cpp \
 *       ../src/DMS.cpp \
 *       ../src/Geodesic.cpp \
 *       ../src/GeodesicLine.cpp
 *
 * See the <a href="Geod.1.html">man page</a> for usage
 * information.
 **********************************************************************/

#include <iostream>
#include <sstream>
#include <string>
#include <sstream>
#include <fstream>
#include <GeographicLib/Geodesic.hpp>
#include <GeographicLib/GeodesicLine.hpp>
#include <GeographicLib/GeodesicExact.hpp>
#include <GeographicLib/GeodesicLineExact.hpp>
#include <GeographicLib/DMS.hpp>
#include <GeographicLib/Utility.hpp>

#if defined(_MSC_VER)
// Squelch warnings about constant conditional expressions and potentially
// uninitialized local variables
#  pragma warning (disable: 4127 4701)
#endif

#include "Geod.usage"

typedef GeographicLib::Math::real real;

std::string LatLonString(real lat, real lon, int prec, bool dms, char dmssep) {
  using namespace GeographicLib;
  return dms ?
    DMS::Encode(lat, prec + 5, DMS::LATITUDE, dmssep) + " " +
    DMS::Encode(lon, prec + 5, DMS::LONGITUDE, dmssep) :
    DMS::Encode(lat, prec + 5, DMS::NUMBER) + " " +
    DMS::Encode(lon, prec + 5, DMS::NUMBER);
}

std::string AzimuthString(real azi, int prec, bool dms, char dmssep) {
  using namespace GeographicLib;
  return dms ? DMS::Encode(azi, prec + 5, DMS::AZIMUTH, dmssep) :
    DMS::Encode(azi >= 180 ? azi - 360 : azi, prec + 5, DMS::NUMBER);
}

std::string DistanceStrings(real s12, real a12,
                            bool full, bool arcmode, int prec, bool dms) {
  using namespace GeographicLib;
  std::string s;
  if (full || !arcmode)
    s += Utility::str<real>(s12, prec);
  if (full)
    s += " ";
  if (full || arcmode)
    s += DMS::Encode(a12, prec + 5, dms ? DMS::NONE : DMS::NUMBER);
  return s;
}

real ReadDistance(const std::string& s, bool arcmode) {
  using namespace GeographicLib;
  return arcmode ? DMS::DecodeAngle(s) : Utility::num<real>(s);
}

int main(int argc, char* argv[]) {
  try {
    using namespace GeographicLib;
    bool linecalc = false, inverse = false, arcmode = false,
      dms = false, full = false, exact = false;
    real
      a = Constants::WGS84_a<real>(),
      f = Constants::WGS84_f<real>();
    real lat1, lon1, azi1, lat2, lon2, azi2, s12, m12, a12, M12, M21, S12;
    real azi2sense = 0;
    int prec = 3;
    std::string istring, ifile, ofile, cdelim;
    char lsep = ';', dmssep = char(0);

    for (int m = 1; m < argc; ++m) {
      std::string arg(argv[m]);
      if (arg == "-i") {
        inverse = true;
        linecalc = false;
      } else if (arg == "-a")
        arcmode = true;
      else if (arg == "-l") {
        inverse = false;
        linecalc = true;
        if (m + 3 >= argc) return usage(1, true);
        try {
          DMS::DecodeLatLon(std::string(argv[m + 1]), std::string(argv[m + 2]),
                            lat1, lon1);
          azi1 = DMS::DecodeAzimuth(std::string(argv[m + 3]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding arguments of -l: " << e.what() << "\n";
          return 1;
        }
        m += 3;
      } else if (arg == "-e") {
        if (m + 2 >= argc) return usage(1, true);
        try {
          a = Utility::num<real>(std::string(argv[m + 1]));
          f = Utility::fract<real>(std::string(argv[m + 2]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding arguments of -e: " << e.what() << "\n";
          return 1;
        }
        m += 2;
      }
      else if (arg == "-d") {
        dms = true;
        dmssep = '\0';
      } else if (arg == "-:") {
        dms = true;
        dmssep = ':';
      } else if (arg == "-b")
        azi2sense = 180;
      else if (arg == "-f")
        full = true;
      else if (arg == "-p") {
        if (++m == argc) return usage(1, true);
        try {
          prec = Utility::num<int>(std::string(argv[m]));
        }
        catch (const std::exception&) {
          std::cerr << "Precision " << argv[m] << " is not a number\n";
          return 1;
        }
      } else if (arg == "-E")
        exact = true;
      else if (arg == "--input-string") {
        if (++m == argc) return usage(1, true);
        istring = argv[m];
      } else if (arg == "--input-file") {
        if (++m == argc) return usage(1, true);
        ifile = argv[m];
      } else if (arg == "--output-file") {
        if (++m == argc) return usage(1, true);
        ofile = argv[m];
      } else if (arg == "--line-separator") {
        if (++m == argc) return usage(1, true);
        if (std::string(argv[m]).size() != 1) {
          std::cerr << "Line separator must be a single character\n";
          return 1;
        }
        lsep = argv[m][0];
      } else if (arg == "--comment-delimiter") {
        if (++m == argc) return usage(1, true);
        cdelim = argv[m];
      } else if (arg == "--version") {
        std::cout
          << argv[0] << ": GeographicLib version "
          << GEOGRAPHICLIB_VERSION_STRING << "\n";
        return 0;
      } else
        return usage(!(arg == "-h" || arg == "--help"), arg != "--help");
    }

    if (!ifile.empty() && !istring.empty()) {
      std::cerr << "Cannot specify --input-string and --input-file together\n";
      return 1;
    }
    if (ifile == "-") ifile.clear();
    std::ifstream infile;
    std::istringstream instring;
    if (!ifile.empty()) {
      infile.open(ifile.c_str());
      if (!infile.is_open()) {
        std::cerr << "Cannot open " << ifile << " for reading\n";
        return 1;
      }
    } else if (!istring.empty()) {
      std::string::size_type m = 0;
      while (true) {
        m = istring.find(lsep, m);
        if (m == std::string::npos)
          break;
        istring[m] = '\n';
      }
      instring.str(istring);
    }
    std::istream* input = !ifile.empty() ? &infile :
      (!istring.empty() ? &instring : &std::cin);

    std::ofstream outfile;
    if (ofile == "-") ofile.clear();
    if (!ofile.empty()) {
      outfile.open(ofile.c_str());
      if (!outfile.is_open()) {
        std::cerr << "Cannot open " << ofile << " for writing\n";
        return 1;
      }
    }
    std::ostream* output = !ofile.empty() ? &outfile : &std::cout;

    const Geodesic geod(a, f);
    const GeodesicExact geode(a, f);
    GeodesicLine l;
    GeodesicLineExact le;
    if (linecalc) {
      if (exact)
        le = geode.Line(lat1, lon1, azi1);
      else
        l = geod.Line(lat1, lon1, azi1);
    }

    // Max precision = 10: 0.1 nm in distance, 10^-15 deg (= 0.11 nm),
    // 10^-11 sec (= 0.3 nm).
    prec = std::min(10 + (sizeof(real) > sizeof(double) ? 3 : 0),
                    std::max(0, prec));
    std::string s;
    int retval = 0;
    while (std::getline(*input, s)) {
      try {
        std::string eol("\n");
        if (!cdelim.empty()) {
          std::string::size_type m = s.find(cdelim);
          if (m != std::string::npos) {
            eol = " " + s.substr(m) + "\n";
            s = s.substr(0, m);
          }
        }
        std::istringstream str(s);
        if (inverse) {
          std::string slat1, slon1, slat2, slon2;
          if (!(str >> slat1 >> slon1 >> slat2 >> slon2))
            throw GeographicErr("Incomplete input: " + s);
          std::string strc;
          if (str >> strc)
            throw GeographicErr("Extraneous input: " + strc);
          DMS::DecodeLatLon(slat1, slon1, lat1, lon1);
          DMS::DecodeLatLon(slat2, slon2, lat2, lon2);
          a12 = exact ?
            geode.Inverse(lat1, lon1, lat2, lon2, s12, azi1, azi2,
                          m12, M12, M21, S12) :
            geod.Inverse(lat1, lon1, lat2, lon2, s12, azi1, azi2,
                         m12, M12, M21, S12);
          if (full)
            *output << LatLonString(lat1, lon1, prec, dms, dmssep) << " ";
          *output << AzimuthString(azi1, prec, dms, dmssep) << " ";
          if (full)
            *output << LatLonString(lat2, lon2, prec, dms, dmssep) << " ";
          *output << AzimuthString(azi2 + azi2sense, prec, dms, dmssep) << " "
                  << DistanceStrings(s12, a12, full, arcmode, prec, dms);
          if (full)
            *output << " " << Utility::str<real>(m12, prec)
                    << " " << Utility::str<real>(M12, prec+7)
                    << " " << Utility::str<real>(M21, prec+7)
                    << " " << Utility::str<real>(S12, std::max(prec-7, 0));
          *output << eol;
        } else {
          if (linecalc) {
            std::string ss12;
            if (!(str >> ss12))
              throw GeographicErr("Incomplete input: " + s);
            std::string strc;
            if (str >> strc)
              throw GeographicErr("Extraneous input: " + strc);
            s12 = ReadDistance(ss12, arcmode);
            if (arcmode)
              exact ?
                le.ArcPosition(s12, lat2, lon2, azi2, a12, m12, M12, M21, S12) :
                l.ArcPosition(s12, lat2, lon2, azi2, a12, m12, M12, M21, S12);
            else
              a12 = exact ?
                le.Position(s12, lat2, lon2, azi2, m12, M12, M21, S12) :
                l.Position(s12, lat2, lon2, azi2, m12, M12, M21, S12);
          } else {
            std::string slat1, slon1, sazi1, ss12;
            if (!(str >> slat1 >> slon1 >> sazi1 >> ss12))
              throw GeographicErr("Incomplete input: " + s);
            std::string strc;
            if (str >> strc)
              throw GeographicErr("Extraneous input: " + strc);
            DMS::DecodeLatLon(slat1, slon1, lat1, lon1);
            azi1 = DMS::DecodeAzimuth(sazi1);
            s12 = ReadDistance(ss12, arcmode);
            if (arcmode)
              exact ?
                geode.ArcDirect(lat1, lon1, azi1, s12, lat2, lon2, azi2, a12,
                                m12, M12, M21, S12) :
                geod.ArcDirect(lat1, lon1, azi1, s12, lat2, lon2, azi2, a12,
                               m12, M12, M21, S12);
            else
              a12 = exact ?
                geode.Direct(lat1, lon1, azi1, s12, lat2, lon2, azi2,
                             m12, M12, M21, S12) :
                geod.Direct(lat1, lon1, azi1, s12, lat2, lon2, azi2,
                            m12, M12, M21, S12);
          }
          if (arcmode)
            std::swap(s12, a12);
          if (full)
            *output << LatLonString(lat1, lon1, prec, dms, dmssep) << " "
                    << AzimuthString(azi1, prec, dms, dmssep) << " ";
          *output << LatLonString(lat2, lon2, prec, dms, dmssep) << " "
                  << AzimuthString(azi2 + azi2sense, prec, dms, dmssep);
          if (full)
            *output << " "
                    << DistanceStrings(s12, a12, full, arcmode, prec, dms)
                    << " " << Utility::str<real>(m12, prec)
                    << " " << Utility::str<real>(M12, prec+7)
                    << " " << Utility::str<real>(M21, prec+7)
                    << " " << Utility::str<real>(S12, std::max(prec-7, 0));
          *output << eol;
        }
      }
      catch (const std::exception& e) {
        // Write error message cout so output lines match input lines
        *output << "ERROR: " << e.what() << "\n";
        retval = 1;
      }
    }
    return retval;
  }
  catch (const std::exception& e) {
    std::cerr << "Caught exception: " << e.what() << "\n";
    return 1;
  }
  catch (...) {
    std::cerr << "Caught unknown exception\n";
    return 1;
  }
}
